tests/README.md

```diff
@@ -0,0 +1,1647 @@
+make sensible use of the suite and to run it as often as possible without long waiting times.
+The FIRST principles makes tests reliable for developers and always delivers the same result
+(no flickering tests). Because the tests themselves verify the result, there are no manual 
+steps to verify the tests.
+
+<a name="fast"></a>
+#### Fast
+Keep the unit tests as fast as possible. They will be run multiple times a day 
+to verify the behaviour of all the classes.
+
+The small grade between fast and slow test can be sometimes a bit blurred, 
+however if the tested part of the code opens database connection or operates with 
+files on a real filesystem the test are always slow.
+
+A lot of slow unit tests are usually an indicator of not so well designed component. 
+Because all the code is tightly coupled to the slow parts or operation of a system.   
+
+<a name="isolated"></a>
+#### Isolated
+Unit-Test focus on small junks of code. So called single *unit*. The more code 
+a test involves the more likely it is that the test fails out of unreasonably 
+circumstances.
+
+For example the code which is under test might interact with other code which 
+connects and interacts with a database. The database itself needs an entire host. 
+So, in fact the test depends on that database and the structure as well as the 
+data. If the datasource is shared between developers the result of the tests are
+no longer reliable because of external changes which are out of control for each 
+individual developer.
+
+Good unit tests also don't depend on other unit tests. For example, all unit tests
+depend on each other to safe some time creating expensive objects. 
+After some time something goes wrong, there will be a massive amount of time 
+spend to found the actual cause because everything is failing due to the 
+high coupling between each test.
+
+Therefore, unit test must be executable any time in any possible order.
+
+The Single Responsibility Principle (SRP) of the SOLID class design principle 
+describes that class should only have one reason to change. This principe is also 
+really good for unit tests because if a test can break for more than one
+reason. It's the best to split the test in multiple cases. 
+"When a focused unit test breaks, it's usually obvious why." [1, Chap. 5]
+
+<a name="repeatable"></a>
+#### Repeatable
+A repeatable test is one which creates the same results all the time. In order to 
+accomplish that, the test must be *isolated*. Each system will interact with elements 
+which are not under the control of the developers. For example, if a system has to 
+deal with dates or time. That means this test have to deal with additional problem 
+which makes writing them more difficult.
+
+In such situations mock objects are used to isolate the class from the outer world. 
+If the dependencies are not mockable there is usually something wrong with the 
+design of the component.
```

Comment:
Maybe an example would help here:
"Most of code in ILIAS uses the global [$DIC](#link-to-some-documentation) e.g. to access the database.
```php
public function insertSomeData($data)
{
            global $DIC;
            $ilDB = $DIC['ilDB'];
            $ilDB->insert(...);
}
```

To test this method you would have to mock the $DIC for every single unit test.
To avoid this the database object could be added via [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection) in the constructor:

```php
public function __construct($database = null)
{
    if($database === null) {
        global $DIC;
        $database = $DIC['ilDB'];
    }
    $this->database = $database;
}

public function insertSomeData($someData)
{
     $this->database->insert(...);
}
```
With example above you can mock every use of the database in  the test, without the need of a `$DIC`